models/article.go

```diff
@@ -0,0 +1,209 @@
+	UpdatedAt      time.Time
+}
+
+type ValidationMessages map[string]interface{}
+
+// NewArticle returns a new Article instance.
+func NewArticle(title string, description string, body string, user *User) *Article {
+	return &Article{
+		Title:       title,
+		Description: description,
+		Body:        body,
+		User:        *user,
+	}
+}
+
+// IsValid check if the article has a valid title, description and body
+func (a *Article) IsValid() (bool, map[string]interface{}) {
+	var errs = ValidationMessages{}
+	var valid = true
+
+	if a.Title == "" {
+		errs["title"] = []string{"title field can't be blank"}
+		valid = false
+	}
+
+	if a.Description == "" {
+		errs["description"] = []string{"description field can't be blank"}
+		valid = false
+	}
+
+	if a.Body == "" {
+		errs["body"] = []string{"body field can't be blank"}
+		valid = false
+	}
+
+	return valid, errs
+}
+
+func (a *Article) IsOwnedBy(username string) bool {
+	return a.User.Username == username
+}
+
+// CreateArticle persist a new article
+func (db *DB) CreateArticle(article *Article) (err error) {
+	err = db.Create(&article).Error
+	return
+}
+
+// DeleteArticle persist a new article
+func (db *DB) DeleteArticle(article *Article) (err error) {
+	err = db.Delete(&article).Error
+	return
+}
+
+func (db *DB) SaveArticle(article *Article) (err error) {
+	err = db.Save(&article).Error
+	return
+}
+
+// GetArticle retrieve an article by it slug
+func (db *DB) GetArticle(slug string) (*Article, error) {
+	var article Article
+	err := db.DB.Scopes(defaultScope).First(&article, "slug = ?", slug).Error
+	return &article, err
+}
+
+// GetAllArticles returns all articles.
+func (db *DB) GetAllArticles() *gorm.DB {
+	return db.DB.Scopes(defaultScope)
+}
+
+func (db *DB) GetAllArticlesWithTag(tagName string) (articles []Article, err error) {
+	//tag := Tag{Name: tagName}
+	//db.FindTag(&tag)
+	err = db. //DB.Model(&tag).
+			Scopes(defaultScope).
+			Joins("JOIN taggings ON taggings.article_id = articles.id").
+			Joins("JOIN tags ON tags.id = taggings.tag_id").
+			Where("tags.name = ?", tagName).
+			Find(&articles).Error
+	//Related(&articles, "Articles").Error
+	return
+}
+
+func (db *DB) GetAllArticlesAuthoredBy(username string) (articles []Article, err error) {
+	//user := User{Username: username}
+	//db.FindUserByUsername(string) *Use
+	err = db.Scopes(defaultScope).
+		Joins("JOIN users ON users.id = articles.user_id").
+		Where("users.username = ?", username).
+		Find(&articles).Error
+	return
+}
+
+func (db *DB) GetAllArticlesFavoritedBy(username string) (articles []Article, err error) {
+	err = db.Scopes(defaultScope).
+		Joins("JOIN favorites ON favorites.article_id = articles.id").
+		Joins("JOIN users ON users.id = favorites.user_id").
+		Where("users.username = ?", username).
+		Find(&articles).Error
+	return
+}
+
+func (db *DB) IsFavorited(userID int, articleID int) bool {
+	f := Favorite{ArticleID: articleID, UserID: userID}
+	if db.Where(f).First(&f).RecordNotFound() {
```

Comment:
Found it when reading the unit tests in the gorm repo. (the code its more helpful than the doc)